### ISSUE TITLE
Prepared Datasets for download

### ISSUE THREAD

**Issue by rohit12** [Bartzi/see#6]
I am unable to find the prepared datasets on the site that you have [linked](https://bartzi.de/research/see#) to. Do I need to prepare the datasets myself? Is it possible to get them in the format suitable for the network?

**Comment by Bartzi** [comment]
Do you want to download the SVHN datasets? If so you can find them (here)[https://bartzi.de/documents/attachment/download?hash_value=8f85bb4379f0974952a826dac2baf6a7_10]

**Comment by rohit12** [comment]
Thank you! This is what I was looking for. One question, though, what is the difference between the "centred" and the "easy" images? Which ones would I use for training the network myself?

One more question: I have a dataset of around 100-500 documents. I wanted to test this network on those documents. How would I go about doing this? Would it be possible for me to get the actual coordinates of the bounding boxes that the network detects?

**Comment by rohit12** [comment]
I am a bit confused by what is mentioned in the dataset part of the README. I have a few questions about it:

> For training you will need:
the file svhn_char_map.json (you can find it in the folder datasets/svhn)
the ground truth files of the dataset you want to use

1. It is mentioned that you would need the ground truth files of the dataset that you are working with. What are the ground truth files in case of the SVHN dataset? I just see the "train" and "validate" folders which contain the images. Are they considered ground truth?

> Add one line to the beginning of each ground truth file: <number of house numbers in image> <max number of chars per house number> (both values need to be separated by a tab character). If you are using the grid dataset it could look like that: 4 4.
In the event that I had to prepare the SVHN dataset on my own, where would I add this? I just see the .png files. I am not really sure about this line. Could you please explain a bit more about this?

Apologies for firing so many questions!

Thanks again, Christian!


>

**Comment by Bartzi** [comment]
We created two different datasets:
1. `easy` contains all images in a predefined grid (figure 5 left in the paper)
2. `centered` contains images where the SVHN numbers are placed randomly in the image, but rather near the center (figure 5 right in the paper)

Regarding your dataset:
It highly depends on what your documents look like. If they are quite similar to the ones you trained the network with (means mainly same classes and similar appearance (you can not use a network trained on MNIST for SVHN and vice versa)), then it should work and you might be able to use the evaluation script, although you might nee to adjust it. You can extract the predicted bboxes from the network. They can be found in the generated sampling grids.

The ground truth files are the `csv` files that contain the path to an image + the labels.

> Add one line to the beginning of each ground truth file: {number of house numbers in image} {max number of chars per house number} (both values need to be separated by a tab character). If you are using the grid dataset it could look like that: 4 4.

This is necessary for training under the curriculum. The line you'll need to add is some metadata about the train dataset and tells the program, how the data looks like. It is used to pad data if the curriculum increases the difficulty.
So you will need to add this to each of the `csv` files. If you take a look at the `png` files in the `easy` directory, you can see that there are always 4 house numbers we want to detect and each number has a maximum length of 4 characters, thus you will need to add `4 4`(digits seperated by tabs) as first line to each of the files.

I hope that helps!

**Comment by rohit12** [comment]
Oh, I see! I need to add the metadata to the start of the `png` files. What tool did you use to add the metadata? Can it be done using a library like PIL?

Also, do you have any suggestions on how would I modify this network for a document? Let's say that there are 3 text regions in the document. The number of chars in the region might vary. Is there any other solution for this other than to do it anyway?

I would be trying to extend your approach to get better accuracy for text detection and recognition than current OCR solutions.

**Comment by Bartzi** [comment]
Oh, I may have stated that wrongly: you do not have to add this to the `png` files. You have to add this metadata as the first row of the `csv` files (`train.csv`, `valid.csv`, `test.csv`).

I do not think that you have to modify the network very much. You'll only have to set the parameters for the maximum number of text regions (that is the number of steps the localization network runs) and the maximum number of characters per time step. So if you have three text regions at every time, you can set this to three, and if the number of characters per text region varies, you have to find a reasonable maximum (lets say its 23) and add this line: `3 23` (separated by a tab) to the beginning of your `train.csv` file. You will also need to pad all your words with the blank label (you can choose which label this should be, but I'd recommend `0` as this is the default)

**Comment by rohit12** [comment]
Hi Christian,

I was able to solve the error. However, I am unable to get NCCL to run properly. It gives me the following error: `Exception: NCCL is not enabled. MultiprocessParallelUpdater requires NCCL.` I have installed NCCL and reinstalled chainer and cupy.

I think I'll have to rewrite your code again. Do you have any suggestions for this?

**Comment by Bartzi** [comment]
Yeah you have two options:

1. try to figure out why NCCL is not properly detected while installing cupy (you could try to reinstall cupy with `pip install --upgrade --force-reinstall --no-cache-dir -v cupy` and look at the output of the compilation of cupy. Search for all errors regarding NCCL and try to fix them.
2. Change the code. This should also be easy, but you will loose the ability to train on multiple GPUs in parallel. You will need to exchange `MultiprocessParallelUpdater` by `StandardUpdater` (if you are using the script `train_svhn.py` this would be in [this](https://github.com/Bartzi/see/blob/master/chainer/train_svhn.py#L146) line, you'll also need to change [this](https://github.com/Bartzi/see/blob/master/chainer/train_svhn.py#L143) line to `train_iterators = chainer.iterators.MultiprocessIterator(gpu_datasets[0], args.batch_size)`)

It should work then ;)

**Comment by rohit12** [comment]
Option 1 worked! I wonder why it did not work when I was manually installing it. 
However, I am now getting this error. Is there any reason for this? Or is it a problem on my end? Or have I forgotten to mention some parameters?

Traceback (most recent call last):
  File "chainer/train_svhn.py", line 146, in <module>
    updater = MultiprocessParallelUpdater(train_iterators, optimizer, devices=args.gpus)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/updaters/multiprocess_parallel_updater.py", line 121, in __init__
    assert len(iterators) == len(devices)
AssertionError

**Comment by Bartzi** [comment]
I think Option 1 woked for you because `pip` is a very nasty program when it comes to reinstalling stuff. If you already have installed a package and want to reinstall it with just `pip install cupy` it won't do anything, even with the flag `--upgrade` or `--force-reinstall` it won't really do so... that is why I suggested this complex command^^

Regarding your error:
Which command line did you use to start the program?

**Comment by rohit12** [comment]
I used the following command:

`python3 chainer/train_svhn.py --char-map ./data/svhn_char_map.json -b 32 ./data/curriculum.json logs/ --blank-label 0`

`curriculum.json`:Path to the csv files
`logs/`: The logs directory

**Comment by rohit12** [comment]
Oh! I see. Yes, I rectified that error. I thought that it would take the default GPU if there wasn't anyone specified.

I am getting another error. 

/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/updaters/multiprocess_parallel_updater.py:131: UserWarning: optimizer.eps is changed to 1e-08 by MultiprocessParallelUpdater for new batch size.
  format(optimizer.eps))

Exception in main training loop: 
Invalid operation is performed in: Concat (Forward)

Expect: in_types.size > 0
Actual: 0 <= 0
Traceback (most recent call last):
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/trainer.py", line 299, in run
    update()
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/updater.py", line 223, in update
    self.update_core()
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/updaters/multiprocess_parallel_updater.py", line 206, in update_core
    loss = _calc_loss(self._master, batch)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/updaters/multiprocess_parallel_updater.py", line 235, in _calc_loss
    return model(*in_arrays)
  File "/home/rohit_shinde12194/see/chainer/utils/multi_accuracy_classifier.py", line 44, in __call__
    self.y = self.predictor(*x)
  File "/home/rohit_shinde12194/see/chainer/models/svhn.py", line 214, in __call__
    return self.recognition_net(images, h)
  File "/home/rohit_shinde12194/see/chainer/models/svhn.py", line 138, in __call__
    final_lstm_predictions = F.concat(final_lstm_predictions, axis=0)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/functions/array/concat.py", line 90, in concat
    y, = Concat(axis).apply(xs)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/function_node.py", line 230, in apply
    self._check_data_type_forward(in_data)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/function_node.py", line 298, in _check_data_type_forward
    self.check_type_forward(in_type)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/functions/array/concat.py", line 22, in check_type_forward
    type_check.expect(in_types.size() > 0)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/utils/type_check.py", line 524, in expect
    expr.expect()
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/utils/type_check.py", line 482, in expect
    '{0} {1} {2}'.format(left, self.inv, right))
Will finalize trainer extensions and updater before reraising the exception.
Traceback (most recent call last):
  File "chainer/train_svhn.py", line 257, in <module>
    trainer.run()
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/trainer.py", line 313, in run
    six.reraise(*sys.exc_info())
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/six.py", line 693, in reraise
    raise value
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/trainer.py", line 299, in run
    update()
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/updater.py", line 223, in update
    self.update_core()
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/updaters/multiprocess_parallel_updater.py", line 206, in update_core
    loss = _calc_loss(self._master, batch)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/training/updaters/multiprocess_parallel_updater.py", line 235, in _calc_loss
    return model(*in_arrays)
  File "/home/rohit_shinde12194/see/chainer/utils/multi_accuracy_classifier.py", line 44, in __call__
    self.y = self.predictor(*x)
  File "/home/rohit_shinde12194/see/chainer/models/svhn.py", line 214, in __call__
    return self.recognition_net(images, h)
  File "/home/rohit_shinde12194/see/chainer/models/svhn.py", line 138, in __call__
    final_lstm_predictions = F.concat(final_lstm_predictions, axis=0)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/functions/array/concat.py", line 90, in concat
    y, = Concat(axis).apply(xs)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/function_node.py", line 230, in apply
    self._check_data_type_forward(in_data)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/function_node.py", line 298, in _check_data_type_forward
    self.check_type_forward(in_type)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/functions/array/concat.py", line 22, in check_type_forward
    type_check.expect(in_types.size() > 0)
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/utils/type_check.py", line 524, in expect
    expr.expect()
  File "/home/rohit_shinde12194/.local/lib/python3.5/site-packages/chainer/utils/type_check.py", line 482, in expect
    '{0} {1} {2}'.format(left, self.inv, right))
chainer.utils.type_check.InvalidType: 
Invalid operation is performed in: Concat (Forward)

I think that the forward pass is failing. However, I am not sure what is causing it to fail.

**Comment by rohit12** [comment]
I am working on a couple of methods for end to end text detection and recognition which would work well on documents. I think your STN-OCR and SEE are one of the few papers on the topics. 

I am thinking of taking a generic text detector and attaching its output to another text recognizer and then training them end-to-end. However, there needs to be a differentiable connection between them. Do you have any idea how this might be done?

**Comment by rohit12** [comment]
While going through your work, I was wondering why you did not use other text recogniton architectures like CRNN with Attention. Is there any reason for that? I was thinking of trying it out.

**Comment by rohit12** [comment]
I am working on a couple of methods for end to end text detection and recognition which would work well on documents. I think your STN-OCR and SEE are one of the few papers on the topics. 

I have some questions and thoughts on this. Do reply as you see fit!

1. I am thinking of taking a generic text detector and attaching its output to another text recognizer and then training them end-to-end. However, there needs to be a differentiable connection between them. Do you have any idea how this might be done?
2. Why did you use a ResNet? In the paper, you argue that gradient information needs to be strong in order to train the network well. I wanted to substitute a ResNet for a CRNN and do some experiments as well. It might lead to some different insights.
3. I am checking out your code. In the `chainer.models` directory, there are the following files: `fsns.py`, `fsns_resnet.py`, `ic_stn.py`, `svhn.py`, `text_recognition.py`. If I want to swap out the ResNet for a different recognition net, I would do it in `text_recognition.py`, right?

**Comment by Bartzi** [comment]
There seems to be something wrong with your arrays. The datatype is not correct, you should check that.

Regarding your questions:
- STN-OCR and SEE are not the only methods for end-to-end detection and recognition. In the last two months I also saw other methods. But the unique property of STN-OCR and SEE is that those systems do not need any groundtruth for the localization of text regions

1. I also had the same problem and I decided to use a Spatial Transformer Network and crop the localized text regions from the image using the differentiable Image Sampler. So far this is the best way I was able to figure out.
2. Well you can substitute the ResNet by a CRNN, but as I said in the paper: I highly recommend using a ResNet like feature extractor because of the strong gradient information that is retained by the ResNet. So if you go for CRNN, use everything that makes a CRNN a CRNN but exchange the convolutional feature extractor by a ResNet like structure and you should be good to go
3. That depends on the task you want to achieve. The names of the files are misleading (I apologize for that). Everything that has to do with `SVHN` experiments can be found in `svhn.py` (that includes the detection and the recognition net for SVHN experiments), everythin that has to do with `FSNS` can be found in `fsns.py` and so on.
4. You can code everything in Tensorflow, too. Even Spatialtransformerrid works there. If you want to do that it should be possible.
5. The most important thing that needs to be improved is the training of the localization network and also the number of maximum localizations. Right now this is too constrained and difficult to learn. A different approach is necessary. In my point of view, this is one of the most important things.

**Comment by rohit12** [comment]
Thanks for the reply, Christian!

How would I go about checking the arrays? I am a bit stumped here, so if you could give me a starting point, I could start my debugging from there.

**Comment by Bartzi** [comment]
this call 

> File "/home/rohit_shinde12194/see/chainer/models/svhn.py", line 138, in call
final_lstm_predictions = F.concat(final_lstm_predictions, axis=0)

is the cause of the problem. So just set a breakpoint in this line and have a look at the list `final_lstm_predictions` there should be something wrong with this array.

**Comment by rohit12** [comment]
I did not get time to look at this yesterday, so I'll be looking at it today. What is a good way to understand your code? Where is the very first layer of the network? I am assuming that this [line](https://github.com/Bartzi/see/blob/master/chainer/models/svhn.py#L41)  is where the model takes in input data. Am I right? I plan on actually going step by step through the models to understand what is going on underneath.

**Comment by Bartzi** [comment]
Yes, you are right. This is the lone where the model takes the input data and does the forward pass through the network.
It might be a good idea to use the python debugger (`pdb` or the ipython debugger `ipdb`) to step through the code and see what is happening where, you should also have a look at the Chainer documentation and learn how a network is built using this library ;)

**Comment by rohit12** [comment]
I finally got to debugging it. This [line](https://github.com/Bartzi/see/blob/master/chainer/models/svhn.py#L121) fails to go into the loop because `num_labels` is 0. When I checked why, I found out that this [line](https://github.com/Bartzi/see/blob/master/chainer/train_svhn.py#L97) gets the number of labels. That command returns a 0. I am not really sure why. Do you have any idea why this happens?

**Comment by Bartzi** [comment]
I'd say that something in your train csv is wrong. could you paste the first 3 lines of the label file you are using for training?

**Comment by rohit12** [comment]
I have pasted them below:

4       4
train/0.png     4       7
train/1.png     5       1
train/2.png     3       3
train/3.png     6       8

**Comment by Bartzi** [comment]
that does not look right... in the first line you are stating, that you want to localize a maximum of 4 text lines per image (first 4) and then, that each text line has a maximum of 4 characters (second 4).
But the next lines only have 2 labels. That is not enough. You should have something like:

```
4 4
train/0.png 4 10 10 10 7 10 10 10 10 10 10 10 10 10 10 10
train/1.png 5 10 10 10 1 10 10 10 10 10 10 10 10 10 10 10
...
```

This assumes that `10` is your blank label.
So make sure to pad your data with the blank label.
For instance the line 

```
train/0.png 4 10 10 10 7 10 10 10 10 10 10 10 10 10 10 10
```

says that you have 2 text lines, the first text line has one character (class 4), the second line has one character (class 7), the next two lines are not there (blank label).

**Comment by rohit12** [comment]
Actually, this is from the data that you linked to in the earlier comments. It is the `train.csv` file from `/data/generated/centered/train.csv`. I am misunderstanding something but what are the blank labels for? Could you give an example with respect to the SVHN dataset?

**Comment by rohit12** [comment]
Oh! Now, I get it. I was misunderstanding it all the while. One more question, I am using the `train.csv` that was present in the SVHN dataset that you linked to. I thought it would  be in the correct format. Wouldn't that `train.csv` work right out of the box?

**Comment by rohit12** [comment]
Your code is running when I use the `train.csv` file present in the `easy` folder. That is the correct format.

One last question, why does your method need to have the number of lines and the number of characters as ground truth? Why can't it infer them? Why can't you let the LSTM run for around 50 timesteps or so?

**Comment by Bartzi** [comment]
Nice, that you got it! I'm sorry it is very difficult to explain properly :sweat_smile:...

The `train.csv` file in the `centered` folder is already in the correct format. The first line of metadata is just missing. This line should be `1 2` because there is always one line of text in the image, and each line of text contains a maximum of 2 characters.

**Comment by Bartzi** [comment]
Sure,

I need the maximum number of lines and maximum number of characters as groundtruth, because of the following reasons:
1. I need to know in advance how many timesteps I have to run my LSTM, because the network has no possibility to know when it should stop, yet
2. Knowing this makes it very easy to create batches, as you always now how many items will be in your batch. This also prevents you from memory exhaustion in the middle of the training, as the memory consumption is constant over every training iteration
3. It is necessary to provide this information in the `csv` file, because training happens under a curriculum (the `json` file you created describes this curriculum)

Does that make sense to you?

**Comment by rohit12** [comment]
1. Yes, I figured that it would be needed for the LSTM. However, other methods for text recognition don't really require the number of timesteps as input (I maybe wrong on this, but I did not notice it). That's why I had this question.


That is my main confusion actually.

**Comment by Bartzi** [comment]
You do not need the number if timesteps as input, you could also do everything without that information in the groundtruth file. It is just necessary to put it into the groundtruth file because of the way I implemented Curriculum Learning. This metadata is necessary, in order to successfully train the model under the curriculum regime. Without the curriculum, there would be no need for this data in the groundtruth file.

**Comment by rohit12** [comment]
Could you explain a bit more? I do not really train with a curriculum. I am just giving the whole `train.csv` file as input and I do not break it up into parts. So how would I do away with that information?

**Comment by Bartzi** [comment]
Since you are not using the curriculum, you could do the following (in `train_*.py`):

1. Build the dataset objects used by the iterators on your own (create two FileBasedDataset objects, one for training and one for validation)
    - make sure to set `file_contains_metadata` to `False`
2. Delete everything that has to do with the curriculum object and exchange the datasets with your FileBasedIterator
3. remove [this](https://github.com/Bartzi/see/blob/master/chainer/train_svhn.py#L244) line

this should give you a basic idea on what you'll need to do in order to get rid of the curriculum.

**Comment by rohit12** [comment]
Thank you! You have answered a lot of my queries.

**Comment by rohit12** [comment]
Hey Christian,

I want to

**Comment by rohit12** [comment]
Hi Christian,

How do you exactly make the curriculum.json file? How do you create multiple training sets of increasing difficulty?

**Comment by rohit12** [comment]
I have a few more questions about the code:

1. What is `fast_validation` and `validation`? What is the difference between the two? How do I plot the `validation` loss and accuracy?
2. Does the above iterator take in the file that I have mentioned in the `curriculum.json` file?

**Comment by rohit12** [comment]
Hi Christian,

Sorry to bother you again! Some other questions that I had:

1. I don't really understand the purpose of `char_map`. How do I create a `char_map` for my own dataset?
2. The `README` is not clear on how to test images. I assume that I'll use `evaluate.py`. However, I do not want to use an evaluation ground truth. I just want to pass a couple of images and see how the network labels them. Could you give me a few pointers regarding this?

Thanks again!

**Comment by Bartzi** [comment]
- `fast_validation` only evaluates a given amount of test iamges (e.g. 200) and not the entire validation set at once. I created this evaluator toget quick evaluation results during an epoch. `validation` is used after each epoch and evaluates the entire validation set.

- the `char_map` is just a mapping from predicted class (i.e. index of classifier) to the character code (unicode) of the character this class represents. Its an easy way of assigning a semantic meaning to the output of a classifier. If you want to create a char_map for your own dataset, you will need to figure out which characters you want your model to predict and then create a `char_map` with these characters, where you assign each character to one output of your classifier.

- Have a look at [this](https://github.com/Bartzi/see#text-recognition-demo) part of the `README` for testing individual images

**Comment by rohit12** [comment]
I am still a bit unclear on how `char_map` corresponds to the unicode. For example, in the `fsns_char_map` you assign `108` to `8364`. Could you explain how this happens? Also, would the network recognize spaces between words as well?

**Comment by rohit12** [comment]
One more thing that I noticed was that in FSNS dataset you do not have any alphabets in the `char_map` for FSNS. How will the network predict alphabets? Have you modified the GT for FSNS such that each alphabet is assigned to a number and when that number is predicted you assume that the alphabet is predicted?

**Comment by Bartzi** [comment]
Alright lets do an example:

Suppose you have samples with the following alphabet (or you want to recognize text with the following alphabet):
`a`, `b`, `c`, `d`. If you now use Python and convert each of these characters to their unicode representation, you will get the following:
```
>>> char_codes = [ord(char) for char in 'abcd']
>>> char_codes
[97, 98, 99, 100]
```
As you are only having an alphabet that consists of 4 characters, you will also need a classifier with 4 classes. That means your `char_map` could look like this:
```
{
    "0": 97,
    "1": 98,
    "2": 99,
    "3": 100
}
```
The char_map is saved as `json` because it is very convenient to do so.
So if you have a look at the my explanations and now think about the example you provided (`108` is mapped to `8364`) you will see that if the network assigns the highest probability to class `108` it thinks that it found the `€` character (this is because `8364` is the unicode representation of the `€` character, try `chr(8364)` in your python shell).

If you also have a closer look at the FSNS char_map you will also notice, that there are all latin characters, indeed.

Regarding your image: I leave creating the char_map up to you as an exercise now.
I would like to shift the discussion about the layout of the groundtruth to your other issue (issue #13)

**Comment by gydlcc** [comment]
what is curriculum learning strategy,

**Comment by mit456** [comment]
@gydlcc `curriculum.json` is a file which contains all paths of `train.csv` and `validation.csv`. Example attached

```python
[{
"train": "/home/Projects/see/datasets/custom/train.csv",
"validation": "/home/Projects/see/datasets/custom/validation.csv"
}]
```
you can add as many dicts as you want and at the time of training you use `curriculum.json` to specify dataset

For e.g in given training command we are passing 
`python train_text_recognition.py --blank-label 0 --char-map ../datasets/custom/custom_char_map.json --gpus 0 --batch-size 10  ../datasets/custom/curriculum.json ../datasets/custom/logs`

**Comment by gydlcc** [comment]
Ths, so where/which directory the .png images shuld i put to train

**Comment by mit456** [comment]
You can keep images where ever you like but you have to mention the absolute path of the image in `train.csv`. so each row in `train.csv` has an absolute path of image and ground truth of that image.

**Comment by gydlcc** [comment]
Ths  pretty much

**Comment by yxnyxnyxnyxn** [comment]
> * `fast_validation` only evaluates a given amount of test iamges (e.g. 200) and not the entire validation set at once. I created this evaluator toget quick evaluation results during an epoch. `validation` is used after each epoch and evaluates the entire validation set.
> * the `char_map` is just a mapping from predicted class (i.e. index of classifier) to the character code (unicode) of the character this class represents. Its an easy way of assigning a semantic meaning to the output of a classifier. If you want to create a char_map for your own dataset, you will need to figure out which characters you want your model to predict and then create a `char_map` with these characters, where you assign each character to one output of your classifier.
> * Have a look at [this](https://github.com/Bartzi/see#text-recognition-demo) part of the `README` for testing individual images

I am confused about the parameter "<path to dataset directory>", I don't know what this parameter should be，Can you help me give an example? thanks a lot

**Comment by Bartzi** [comment]
Yeah, `path to dataset directory` is bullshit, it should be the path to the log directory... I'm sorry for that I'm going to fix the README accordingly.

**Comment by yxnyxnyxnyxn** [comment]
> Yeah you have two options:
> 
> 1. try to figure out why NCCL is not properly detected while installing cupy (you could try to reinstall cupy with `pip install --upgrade --force-reinstall --no-cache-dir -v cupy` and look at the output of the compilation of cupy. Search for all errors regarding NCCL and try to fix them.
> 2. Change the code. This should also be easy, but you will loose the ability to train on multiple GPUs in parallel. You will need to exchange `MultiprocessParallelUpdater` by `StandardUpdater` (if you are using the script `train_svhn.py` this would be in [this](https://github.com/Bartzi/see/blob/master/chainer/train_svhn.py#L146) line, you'll also need to change [this](https://github.com/Bartzi/see/blob/master/chainer/train_svhn.py#L143) line to `train_iterators = chainer.iterators.MultiprocessIterator(gpu_datasets[0], args.batch_size)`)
> 
> It should work then ;)

@Bartzi
I also encountered the same problem when i use the script "train_text_recognition.py", and  I have tried the methods above. But it is not useful for me , the error log is 
"
$ python train_text_recognition.py  /home/xuenan/project/cable_reg/see/src/see/chainer/text_recognition_model/small_dataset/curriculum.json text_recognition_model/log/ --char-map text_recognition_model/small_dataset/ctc_char_map.json --blank-label 0 -b 1
Traceback (most recent call last):
  File "train_text_recognition.py", line 180, in <module>
    converter=get_concat_and_pad_examples(args.blank_label)
  File "/usr/local/lib/python3.5/dist-packages/chainer/training/updaters/multiprocess_parallel_updater.py", line 116, in __init__
    'NCCL is not enabled. MultiprocessParallelUpdater '
Exception: NCCL is not enabled. MultiprocessParallelUpdater requires NCCL.
Please reinstall chainer after you install NCCL.
(see https://github.com/chainer/chainer#installation).
"

**Comment by Bartzi** [comment]
> the meaning of 23 1

I know this is confusing! But for doing the text recognition we are using a trickby changing the semantics of the whole thing. As we want to have a bounding box for each character, we handle each character as being a word and each word only consists of a single character, leading to the first line being `23 1`, meaning we want to find a maximum of 23 words (in our case characters) and each word (each character) has the length `1`.

Does that make it clearer, why it is used for text recogniton in that way?

**Comment by harshalcse** [comment]
@rohit12 
I'm tring to train SVHN dataset using following train.csv which path is defined in curricum.json as follows 
![train](https://user-images.githubusercontent.com/5838252/54102268-b71dbb80-43ed-11e9-8629-1e3be6c38b83.PNG)
Command for trainning svhn dataset is 
sudo python3 chainer/train_svhn.py 
"/root/see-master/datasets/train_dataextract_train/curriculum.json" 0 
--char-map /root/see-master/datasets/svhn/svhn_char_map.json 
--test-image /root/see-master/datasets/test_dataextract/train/0.png 
-b 60

but getting error as follows:
Traceback (most recent call last):
  File "chainer/train_svhn.py", line 75, in <module>
    train_dataset, validation_dataset = curriculum.load_dataset(0)
  File "/root/see-master/chainer/utils/baby_step_curriculum.py", line 38, in load_dataset
    train_dataset = self.dataset_class(self.train_curriculum[level], **self.dataset_args)
  File "/root/see-master/chainer/datasets/file_dataset.py", line 31, in __init__
    self.num_timesteps, self.num_labels = (int(i) for i in next(reader))
  File "/root/see-master/chainer/datasets/file_dataset.py", line 31, in <genexpr>
    self.num_timesteps, self.num_labels = (int(i) for i in next(reader))
ValueError: invalid literal for int() with base 10: '/root/see-master/datasets/train_dataextract_train/train/0.png'

Request you to help me to resolve the issue.

**Comment by Bartzi** [comment]
@harshalcse please stop spamming other issues and open a new issue for your problem next time!

**Comment by harshalcse** [comment]
> Yeah you have two options:
> 
> 1. try to figure out why NCCL is not properly detected while installing cupy (you could try to reinstall cupy with `pip install --upgrade --force-reinstall --no-cache-dir -v cupy` and look at the output of the compilation of cupy. Search for all errors regarding NCCL and try to fix them.
> 2. Change the code. This should also be easy, but you will loose the ability to train on multiple GPUs in parallel. You will need to exchange `MultiprocessParallelUpdater` by `StandardUpdater` (if you are using the script `train_svhn.py` this would be in [this](https://github.com/Bartzi/see/blob/master/chainer/train_svhn.py#L146) line, you'll also need to change [this](https://github.com/Bartzi/see/blob/master/chainer/train_svhn.py#L143) line to `train_iterators = chainer.iterators.MultiprocessIterator(gpu_datasets[0], args.batch_size)`)
> 
> It should work then ;)

@Bartzi I tried with second option but it gives following error to me.
Traceback (most recent call last):
  File "chainer/train_svhn.py", line 146, in <module>
    updater = StandardUpdater(train_iterators, optimizer, devices=args.gpus)
TypeError: __init__() got an unexpected keyword argument 'devices'
